### PR TITLE
fix some bugs, add an asset-uploader with copiable URL

### PR DESCRIPTION
**Pull request by birarda**

### CODE REVIEW

File: interface/src/Menu.cpp [hunk 1]
```diff
@@ -383,7 +353,20 @@ Menu::Menu() {
...
+    
+    MenuWrapper* assetDeveloperMenu = developerMenu->addMenu("Assets");
+    
+    auto& assetDialogFactory = AssetUploadDialogFactory::getInstance();
+    assetDialogFactory.setParent(this);
+    
+    QAction* assetUpload = addActionToQMenuAndActionHash(assetDeveloperMenu,
+                                                         MenuOption::UploadAsset,
+                                                         0,
+                                                         &AssetUploadDialogFactory::getInstance(),
```
Comment (huffman) [inline interface/src/Menu.cpp]: can we use the reference from above?

File: interface/src/ui/AssetUploadDialogFactory.cpp [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+#include <QtWidgets/QDialogButtonBox>
+#include <QtWidgets/QFileDialog>
+#include <QtWidgets/QLabel>
+#include <QtWidgets/QMessageBox>
+#include <QtWidgets/QLineEdit>
+#include <QtWidgets/QVBoxLayout>
+
+AssetUploadDialogFactory& AssetUploadDialogFactory::getInstance() {
+    static AssetUploadDialogFactory staticInstance;
+    return staticInstance;
```
Comment (huffman) [inline interface/src/ui/AssetUploadDialogFactory.cpp:29]: this is cool

File: libraries/networking/src/AssetUpload.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            } else {
+                // error during upload - emit finished with an empty hash
+                // callers can get the error from this object
+                
+                // TODO: get the actual error from the callback
+                _result = PermissionDenied;
+            
+                emit finished(this, hash);
+            }
+        });
```
Comment (huffman) [inline libraries/networking/src/AssetUpload.cpp:59]: We should set an error result and emit finished if we weren't able to open the file

File: interface/src/ui/AssetUploadDialogFactory.cpp [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            }
+        }
+    } else {
+        // we don't have permission to upload to asset server in this domain - show the permission denied error
+        showErrorDialog(QString(), PERMISSION_DENIED_ERROR);
+    }
+    
+}
+
+void AssetUploadDialogFactory::handleUploadFinished(AssetUpload* upload, const QString& hash) {
```
Comment (huffman) [inline interface/src/ui/AssetUploadDialogFactory.cpp:71]: Do we delete the upload object anywhere in here? I think we should call `upload->deleteLater()` at the end

File: interface/src/ui/AssetUploadDialogFactory.cpp [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+                QObject::connect(upload, &AssetUpload::finished, this, &AssetUploadDialogFactory::handleUploadFinished);
+                
+                // start the upload now
+                upload->start();
+            } else {
+                // show a QMessageBox to say that there is no local asset server
+                QString messageBoxText = QString("Could not upload \n\n%1\n\nbecause you are currently not connected" \
+                                                 " to a local asset-server.").arg(QFileInfo(filename.toString()).fileName());
+                
+                QMessageBox::information(_dialogParent, "Failed to Upload", messageBoxText);
```
Comment (huffman) [inline interface/src/ui/AssetUploadDialogFactory.cpp:61]: Looks like we leak `upload`, we should just call `upload->deleteLater()` if we reach this point

Comment (birarda) [inline interface/src/ui/AssetUploadDialogFactory.cpp:61]: it's a nullptr here

File: interface/src/ui/AssetUploadDialogFactory.cpp (hunk 4 above)
Comment (birarda) [inline interface/src/ui/AssetUploadDialogFactory.cpp:71]: added

File: interface/src/ui/AssetUploadDialogFactory.cpp (hunk 5 above)
Comment (huffman) [inline interface/src/ui/AssetUploadDialogFactory.cpp:61]: I'm wrong, `upload` is `nullptr` here

**Pull request merged**